src/consensus/blockstore.rs

```diff
@@ -265,70 +138,105 @@ impl Blockstore {
     /// Returns `None` if blockstore does not hold any version of the block yet.
     #[must_use]
     pub fn canonical_block_hash(&self, slot: Slot) -> Option<Hash> {
-        self.canonical.get(&slot).copied()
+        self.slot_data(slot)?
+            .canonical
+            .completed
+            .clone()
+            .map(|c| c.0)
     }
 
-    /// Gives any relevant alternative block hashes for a given slot, if any.
-    #[must_use]
-    pub fn alternative_block_hashes(&self, slot: Slot) -> Option<&[Hash]> {
-        self.alternatives.get(&slot).map(|v| v.as_ref())
+    /// Gives the number of stored slices for a given `slot`.
+    pub fn stored_slices_for_slot(&self, slot: Slot) -> usize {
+        self.slot_data(slot).map_or(0, |s| s.canonical.slices.len())
+    }
+
+    /// Gives the number of stored shreds for a given `slot` (across all slices).
+    pub fn stored_shreds_for_slot(&self, slot: Slot) -> usize {
+        self.slot_data(slot).map_or(0, |s| {
+            s.canonical.shreds.values().map(Vec::len).sum::<usize>()
+        })
     }
 
+    /// Gives the number of stored shreds for a given slot and slice.
+    pub fn stored_shreds_for_slice(&self, slot: Slot, slice: usize) -> usize {
+        self.slot_data(slot)
+            .map_or(0, |s| s.canonical.shreds.get(&slice).map_or(0, Vec::len))
+    }
+
+    // /// Gives any relevant alternative block hashes for a given slot, if any.
+    // #[must_use]
+    // pub fn alternative_block_hashes(&self, slot: Slot) -> &[Hash] {
+    //     self.slot_data(slot)
+    //         .map_or(&[], |s| s.alternatives.as_ref())
+    // }
+
     /// Gives reference to stored block for the given `slot` and `hash`.
     ///
     /// Returns `None` if blockstore does not hold that block yet.
     pub fn get_block(&self, slot: Slot, hash: Hash) -> Option<&Block> {
-        self.blocks.get(&(slot, hash))
+        let slot_data = self.slot_data(slot)?;
+        if let Some((h, block)) = slot_data.canonical.completed.as_ref() {
+            if *h == hash {
+                return Some(block);
+            }
+        } else {
+            return slot_data
+                .alternatives
+                .get(&hash)
+                .and_then(|data| data.completed.as_ref().map(|(_, block)| Some(block)))
+                .flatten();
+        }
+        return None;
```

Comment:
```suggestion
    match slot_data.canonical.completed.as_ref() {
        Some((h, block)) => {
            if *h == hash {
                return Some(block);
            }
            None
        }
        
        None => {
            slot_data
                .alternatives
                .get(&hash)
                .and_then(|data| data.completed.as_ref().map(|(_, block)| Some(block)))
                .flatten();
        }
    }
```

I somehow find this version a bit easier to read.  It is easier to understand which conditions hold when and what we are returning.

Reply:
this was actually a bug, the logic should be like this: 
```
        if let Some((h, block)) = slot_data.canonical.completed.as_ref() {
            if *h == hash {
                return Some(block);
            }
        }
        slot_data
            .alternatives
            .get(&hash)
            .and_then(|data| data.completed.as_ref().map(|(_, block)| Some(block)))
            .flatten()
```